### DISCUSSION TITLE
MLIR index.cmp

### DISCUSSION THREAD

**Discussion by alfie-nsugh** [modular/modular#1137] [Q&A] [answered]
In the tutorial on low level IR in mojo there is this line:
```
return Self(__mlir_op. `index.cmp`.[pred=`#index<cmp_predicate eq>`](lhsIndex, rhsIndex) 
```
I'm just a bit confused by #index<cmp_predicate, I found the syntax a bit random. I wanted to see if I can make structs as kind of wrappers around other MLIR dialects, but it's hard to generalize the tutorial when the syntax doesn't make sense to you. Can someone explain what the general idea is around the syntax here?

**Comment by lattner** [comment, marked as answer]
The syntax within the backticks is the low-level raw MLIR syntax for the cmp_predicate enum in the index dialect.  This is documented here:
https://mlir.llvm.org/docs/Dialects/IndexOps/#indexcmp-indexcmpop

MLIR isn't super pretty about this, because it often gets sugared, e.g. at the operation level, but this needs to use the explicit syntax for it.

**Reply by alfie-nsugh** [reply]
So I suppose the general pattern would be:
```__mlir_op.`operation`[attribute=__mlir_attr.`#dialect<my_attribute value>`]```
Where
- `__mlir_op` is used to invoke a specific MLIR operation.
- `operation` is the name of the operation you're invoking.
- `[attribute=` specifies an attribute within the operation.
- `__mlir_attr` is used to reference an attribute.
- `#dialect` refers to the specific dialect containing the attribute.
- `<my_attribute value>` is the value you want to use from the dialect.

Thanks for responding by the way, you must be busy.

Edit: I think this link explains the syntax quite well:
https://mlir.llvm.org/docs/LangRef/#dialect-attribute-values
